src/plugins/intel_npu/src/plugin/npuw/compiled_model.cpp

```diff
@@ -393,24 +399,46 @@ ov::npuw::CompiledModel::CompiledModel(const std::shared_ptr<ov::Model>& model,
     reset_io();
 }
 
-void ov::npuw::CompiledModel::fill_weights_bank(const std::size_t idx) {
-    LOG_VERB("Filling weights bank for Subgraph[" << idx << "]...");
-    LOG_BLOCK();
+void ov::npuw::CompiledModel::finalize_weights_bank() {
+    ov::parallel_for(m_compiled_submodels.size(), [&](std::size_t idx) {
+        auto& comp_model_desc = m_compiled_submodels[idx];
+
+        if (!comp_model_desc.replaced_by) {
+            return;
+        }
+
+        const auto real_idx = comp_model_desc.replaced_by.value();
+        auto& func_desc = m_compiled_submodels[real_idx];
 
-    NPUW_ASSERT(m_compiled_submodels[idx].replaced_by);
+        // Due to concat some tensor should be skipped in closure
+        m_compiled_submodels[idx].closure.resize(0);
+        m_compiled_submodels[idx].is_remote.resize(0);
 
-    auto& comp_model_desc = m_compiled_submodels[idx];
+        for (std::size_t tidx = 0; tidx < comp_model_desc.transformations.size(); ++tidx) {
+            const auto& lt = m_compiled_submodels[idx].transformations[tidx];
 
-    for (std::size_t cidx = 0u; cidx < comp_model_desc.closure.size(); cidx++) {
-        comp_model_desc.closure[cidx] = m_weights_bank->update(comp_model_desc.closure[cidx]);
-        if (m_cfg.get<::intel_npu::NPUW_FOLD>()) {
-            comp_model_desc.update_required[cidx] = true;
-        } else {
-            comp_model_desc.update_required[cidx] = false;
+            // FIXME: probably should be more careful with the devices here
+            if (!m_weights_bank->has(lt, *func_desc.device_it)) {
+                ov::Tensor evaled = lt.eval();
+                if (lt.has_concat()) {
+                    // FIXME: probably setting just front() LazyTensor here is enough
+                    for (const auto& lt_to_concat : lt.get_lt_to_concat()) {
+                        // Note: this also includes this LazyTensor's original ov::Tensor
+                        m_weights_bank->store(lt_to_concat, evaled, *func_desc.device_it);
+                    }
+                } else {
+                    m_weights_bank->store(lt, evaled, *func_desc.device_it);
+                }
+            }
+
+            m_compiled_submodels[idx].closure.push_back(m_weights_bank->get(lt, *func_desc.device_it));
+            m_compiled_submodels[idx].is_remote.push_back(true);
```

Comment:
what happens here? why new closures are introduced here?

Reply:
At this point there are no closures in the .closure vector. I'm adding them after evaluating LazyTensor